### DISCUSSION TITLE
[Bug]: Zookeeper Validation Error on Kafka Cluster Update with KRaft Enabled

### DISCUSSION THREAD

**Discussion by yanmxa** [strimzi/strimzi-kafka-operator#10836] [General]
### Description
When attempting to update the Kafka cluster, the following error occurs:

```
Kafka.kafka.strimzi.io "kafka" is invalid: [spec.zookeeper.replicas: Required value, spec.zookeeper.storage: Required value]
```

### Configuration Details

The Kafka resource configuration is as follows:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  annotations:
    strimzi.io/kraft: enabled
    strimzi.io/node-pools: enabled
  creationTimestamp: "2024-11-14T04:16:15Z"
  generation: 2
  labels:
    global-hub.open-cluster-management.io/managed-by: global-hub-operator
  name: kafka
  namespace: multicluster-global-hub
  ownerReferences:
  - apiVersion: operator.open-cluster-management.io/v1alpha4
    blockOwnerDeletion: true
    controller: true
    kind: MulticlusterGlobalHub
    name: multiclusterglobalhub
    uid: 8aeaf2d3-07c4-4923-8284-f9f704dc7d84
  resourceVersion: "47817"
  uid: e692a4a6-86e7-4f2f-afe9-0bdbe0649823
spec:
  entityOperator:
    topicOperator: {}
    userOperator: {}
  kafka:
    authorization:
      type: simple
    config:
      default.replication.factor: 1
      min.insync.replicas: 1
      offsets.topic.replication.factor: 1
      transaction.state.log.min.isr: 1
      transaction.state.log.replication.factor: 1
    listeners:
    - name: plain
      port: 9092
      tls: false
      type: internal
    - authentication:
        type: tls
      configuration:
        bootstrap:
          nodePort: 30095
        brokers:
        - advertisedHost: 172.18.0.2
          broker: 0
      name: external
      port: 9095
      tls: true
      type: nodeport
    replicas: 1
    storage:
      type: ephemeral
    version: 3.7.0
```

### Problem
Although the KRaft mode is enabled in the configuration, the system raises an error related to missing Zookeeper settings (`spec.zookeeper.replicas` and `spec.zookeeper.storage`), which should not be required in KRaft mode.

### Request
Please investigate why Zookeeper configuration is still required when KRaft mode is enabled, or advise on any additional configuration needed to suppress these requirements in KRaft mode.

### Expected behavior

With `strimzi.io/kraft: enabled` annotation, Zookeeper configuration should not be mandatory, as KRaft mode removes the need for Zookeeper.

### Strimzi version

0.41.0

### Kubernetes version

KinD cluster kubernetes version: v1.23.0

**Comment by katheris** [comment]
Hi @yanmxa, your Kafka CR doesn't look quite right. For example you have `replicas` and `storage` included, when those should be in your KafkaNodePool CR(s). What steps did you take before getting the error? Are you updating an existing Kafka cluster?

Here's an example of a valid KRaft cluster: https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/kafka/kraft/kafka-single-node.yaml

**Comment by scholzj** [comment]
It is not really clear where you get the error. If you get it when trying to create/update the resource in Kubernetes, it sounds like you did not update the CRDs.

**Comment by yanmxa** [comment]
When I attempt to update the PR, I encounter the above error. 

Thanks @katheris and @scholzj!  

1. I tried updating the existing Kafka CR without the `replicas` and `storage` fields even though I have configured them in the `KafkaNodePool` CR, but I received a validation error:
    ```
    Kafka.kafka.strimzi.io "kafka" is invalid: [spec.kafka.replicas: Required value, spec.kafka.storage: Required value]
    ```
    So I add the  `replicas` and `storage` in the above CR. 

2. It might be due to the lower CRD version. The CSV I’m currently using is `strimzi-cluster-operator.v0.41.0`. Let me try a higher version. :)

**Reply by scholzj** [reply]
As I said, if you get the error from `kubectl`, it is about the CRDs.

**Reply by yanmxa** [reply]
👍 @scholzj Got it! I tried it with Strimzi CSV 0.42.0, and the error is gone!
